drivers/sensor: extend lps22df driver to also support lps28dfw pressure sensor

jonas-rem opened the pull request:
Add support for the ST LPS28DFW pressure sensor. The sensor differs in the following points from the otherwise similar LPS22DF:

Dual FS (260 to 1060 hPa and 260 to 4060 hPa)
I2C and I3C, no SPI
Waterproof (no SW consequences)

This as a new approach, compared to #65164. It extends the existing driver, instead of creating a new driver.

Related Issues/PRs:
- Feature Request issue: https://github.com/zephyrproject-rtos/zephyr/issues/64766
- Previous PR: https://github.com/zephyrproject-rtos/zephyr/pull/65164

~~**Note:** This PR is WIP for now to make commenting easier. Currently the chip_api is not complete and some parts will still change significantly. I know there are some issues and review comments that I have not yet addressed. I'll work on it and will set it from draft to PR in the next days.
@fabiobaltieri @avisconti~~

Commit:
drivers/sensor: add support to LPS28DFW pressure sensor

The LPS28DFW is an ultracompact, piezoresistive, absolute pressure sensor.
Compared to the LPS22DF, the LPS28DFW is waterproof and has a Dual FS
capability and does not have SPI. This commit extends the LPS22DF driver to
be compatible with the LPS28DFW device.

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

fabiobaltieri commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,352 @@
+#endif /* CONFIG_LPS2XDF_TRIGGER */
+
+#define LPS2XDF_CONFIG_COMMON(inst, name)                                                          \
+	.odr = DT_INST_PROP(inst, odr), .lpf = DT_INST_PROP(inst, lpf),                            \
+	.avg = DT_INST_PROP(inst, avg), .drdy_pulsed = DT_INST_PROP(inst, drdy_pulsed),            \
+	.chip_api = &name##_chip_api,                                                              \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, drdy_gpios), (LPS2XDF_CFG_IRQ(inst)), ())
```
```c
        .odr = DT_INST_PROP(inst, odr),                                                             
        .lpf = DT_INST_PROP(inst, lpf),                                                             
        .avg = DT_INST_PROP(inst, avg),                                                             
        .drdy_pulsed = DT_INST_PROP(inst, drdy_pulsed),                                             
        .chip_api = &name##_chip_api,                                                               
        IF_ENABLED(DT_INST_NODE_HAS_PROP(inst, drdy_gpios), (LPS2XDF_CFG_IRQ(inst)))                
```

fabiobaltieri commented on drivers/sensor/lps2xdf/lps2xdf.h:
```diff
@@ -0,0 +1,106 @@
+		const struct i3c_device_id dev_id;
+	} i3c;
+#endif
+	const void *chip_api;
```
you can use `lps2xdf_chip_api` here, it's not like you are making this generic

jremmert-phytec-iot replied:
Ok, changed it.

fabiobaltieri commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,352 @@
+	const struct lps2xdf_config *const cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lps22df_md_t md;
+	struct lps2xdf_chip_api *chip_api = (struct lps2xdf_chip_api *)cfg->chip_api;
```
don't need the casting here

jremmert-phytec-iot commented on drivers/sensor/lps2xdf/lps2xdf_trigger.c:
```diff
@@ -232,7 +234,8 @@ int lps22df_init_interrupt(const struct device *dev)
 	}
 
-#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i3c)
+#if (DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(st_lps22df, i3c) || \
+	DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(st_lps28dfw, i3c))
```
This would require whitespaces at the start of the line and checkpatch will complain about it. Do you know any workaround for that?

fabiobaltieri replied:
Ah... how have I never bumped into that?! Alright ignore my comment  :-)

jremmert-phytec-iot commented on dts/bindings/sensor/st,lps28dfw-common.yaml:
```diff
@@ -0,0 +1,23 @@
+        Specify the full-scale mode.
+        The default is the power-on reset value.
+
+        0 = Mode 0, full scale up to 1260 hPa
+        1 = Mode 1, full scale up to 4060 hPa
+    enum:
+      - 0
+      - 1
```
Right, did not see that PR. I made those defines agnostic to the two sensors, except `fs` (Full-Scale Mode) which can only be set for the lps28dfw device. Those changes are in a separate commit.

jremmert-phytec-iot commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,352 @@
+
+#define LPS2XDF_DEFINE(inst, name)                                                                 \
+	static struct lps2xdf_data lps2xdf_data_##name##_##inst;                                   \
+	static const struct lps2xdf_config lps2xdf_config_##name##_##inst = COND_CODE_1(           \
+		DT_INST_ON_BUS(inst, spi), (LPS2XDF_CONFIG_SPI(inst, name)),                       \
+		(COND_CODE_1(DT_INST_ON_BUS(inst, i3c), (LPS2XDF_CONFIG_I3C_OR_I2C(inst, name)),   \
```
Yes, much better, changed it.

jremmert-phytec-iot commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,352 @@
+			LPS2XDF_CONFIG_COMMON(inst, name)                                          \
+	}
+
+/*
+ * Instantiation macros used when a device is on an I2C bus.
+ */
```
dropped

jremmert-phytec-iot commented:
Thanks for your review @fabiobaltieri! I changed most of your annotations, also the one from the previous PR.

The driver is feature-complete now. It looks like quite a bit overhead for me, let me know if you have ideas how to simplify it.

The test that I used with the reel boad and two lps28dfw sensors is not included anymore. But for reference: https://github.com/jremmert-phytec-iot/zephyr/commit/eed0ff15bac8b94c7c79de663b5eb3af34d6d511

fabiobaltieri commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,491 @@
+
+#define LPS2XDF_CONFIG_I3C(inst, name)                                                             \
+{                                                                                                  \
+	STMEMSC_CTX_I3C(&lps2xdf_config_##name##_##inst.stmemsc_cfg),                              \
+		.stmemsc_cfg =                                                                     \
+			{                                                                          \
+				.i3c = &lps2xdf_data_##name##_##inst.i3c_dev,                      \
+			},                                                                         \
+		.i3c.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),                                       \
+		.i3c.dev_id = I3C_DEVICE_ID_DT_INST(inst),                                         \
+			      LPS2XDF_CONFIG_COMMON(inst, name)                                    \
```
should prob be something like

```c
        STMEMSC_CTX_I3C(&lps2xdf_config_##name##_##inst.stmemsc_cfg),                              \
        .stmemsc_cfg = {                                                                           \
                .i3c = &lps2xdf_data_##name##_##inst.i3c_dev,                                      \
        },                                                                                         \
        .i3c.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),                                               \
        .i3c.dev_id = I3C_DEVICE_ID_DT_INST(inst),                                                 \
        LPS2XDF_CONFIG_COMMON(inst, name) 
```

same for the two above, the current alignment is confusing and clang-format makes a mess as usual :-)

fabiobaltieri commented:
> The driver is feature-complete now. It looks like quite a bit overhead for me, let me know if you have ideas how to simplify it.

I think it's alright, what's the overhead concern specifically?

> The test that I used with the reel boad and two lps28dfw sensors is not included anymore. But for reference: [jremmert-phytec-iot@eed0ff1](https://github.com/jremmert-phytec-iot/zephyr/commit/eed0ff15bac8b94c7c79de663b5eb3af34d6d511)

Thanks, was wondering about it :-)

jremmert-phytec-iot commented:
> > The driver is feature-complete now. It looks like quite a bit overhead for me, let me know if you have ideas how to simplify it.
> 
> I think it's alright, what's the overhead concern specifically?

The main concern was the flash occupied by the driver.

Example: Hello world with `CONFIG_SENSOR=y`, `CONFIG_LOG=n`:
- no pressure sensors in dts: 56380 B
- one pressure sensor: 59766 B
- two of the same pressure sensors (`lps28dfw` + `lps28dfw`): 59840 B
- two different pressure sensors (`lps22df` + `lps28dfw`): 61100 B

The first device adds more than 3 KB, I felt this is more than for the average sensor driver.


The second concern is general code complexity for a driver. which is significantly higher than a stand-alone driver. But if the approach has the potential to add more ST Pressure sensor variants it would probably justify  a higher complexity.

> 
> > The test that I used with the reel boad and two lps28dfw sensors is not included anymore. But for reference: [jremmert-phytec-iot@eed0ff1](https://github.com/jremmert-phytec-iot/zephyr/commit/eed0ff15bac8b94c7c79de663b5eb3af34d6d511)
> 
> Thanks, was wondering about it :-)
> 
> On a more general comment, I think in retrospective it would have been good to have two different commits: one to rename the file and functions, one to add the extra API. Then it would have been easier to review the split part alone, the diff now looks pretty much like an implementation from scratch, which makes me see all the nitpick opportunities missed in the first review :-)

True, I could still separate it when I'm addressing your comments. That would simplify the work of additional reviewers at least. But before merging we have to squash it, right?

fabiobaltieri commented:
> The main concern is the flash occupied by the driver.
> 
> Example: Hello world with `CONFIG_SENSOR=y`, `CONFIG_LOG=n`:
> 
>     * no pressure sensors in dts: 56380 B
> 
>     * one pressure sensor: 59766 B
> 
>     * two of the same pressure sensors (`lps28dfw` + `lps28dfw`): 59840 B
> 
>     * two different pressure sensors (`lps22df` + `lps28dfw`): 61100 B
> 
> 
> The first device adds more than 3 KB, I felt this is more than for the average sensor driver.
> 
> The second concern is general code complexity for a driver, which is significantly higher than a stand-alone driver. But if the approach has the potential to add more ST Pressure sensor variants, it would probably justify a higher complexity.

Yeah well, that's the compromise for having a common code base + HAL I'm afraid. The "two different type of sensors" seems like a bit of an artificial case anyway, I'd imagine an actual application to use only one if they need multiple channels.

If you want to dig further, try `west build -t rom_report`, is it really the driver that takes space or is it having multiple hals built in?

> True, I could still separate it when I'm addressing your comments. That would simplify the work of additional reviewers at least. But before merging we have to squash it, right?

No we don't, what you are doing is right (addressing comments in the original stack and force pushing). We don't squash on merge, leaving comments addressing things would just make a giant mess for reviewer stepping in half way, just force push without changing base if possible and it's all good.

Commit:
dt-bindings: sensor: lps2xdf: make macros sensor agnostic

The possible settings for the lps22df and lps28dfw sensor are overlapping,
except the Full-Scale mode setting. Threrefore we can change the macros
from LPS22DF_DT_[..] to LPS2xDF_DT_[..].

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

jremmert-phytec-iot commented:
When looking closer at the flash consumption of individual components, things look ok to me.

Device HAL for each device, e.g. lps22df_STdC: ~900 B
Device Driver for two devices, lps2xdf: ~1400 B, one device ~900 B

avisconti commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,490 @@
+	}
+
+#ifdef CONFIG_LPS2XDF_TRIGGER
+	if (lps2xdf_init_interrupt(dev, DEVICE_VARIANT_LPS28DFW) < 0) {
```
Actually I think this code is wrong. It was wrong also in lps22df (my bad). I'm starting to review the code for all ST sensors and since we are changing this one we can do the change on thi PR itself. Basically here the problem is that lps22df can be multi instantiated, with one instance with triggers and one with not. So, we do not have to rely on ```CONFIG_LPS2XDF_TRIGGER```, but instead on the DT (```DT_INST_NODE_HAS_PROP``` for drdy_gpios). See lsm6dsv16x implementation, where .trig_enabled is set to '1' if either one of int gpios is defined in DT.

jremmert-phytec-iot replied:
Right, I did not see that. It's changed now.

avisconti commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,490 @@
+
+
+#ifdef CONFIG_LPS2XDF_TRIGGER
+#define LPS2XDF_CFG_IRQ(inst) .gpio_int = GPIO_DT_SPEC_INST_GET(inst, drdy_gpios),
```
here we need to define trig_enabled (see what I said before)

jremmert-phytec-iot commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,490 @@
+	.odr = DT_INST_PROP(inst, odr),                                        \
+	.lpf = DT_INST_PROP(inst, lpf),                                        \
+	.avg = DT_INST_PROP(inst, avg),                                        \
+	.drdy_pulsed = DT_INST_PROP(inst, drdy_pulsed),                        \
```
moved

jremmert-phytec-iot commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,490 @@
+}
+#endif
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_lps28dfw)
```
Good point, this makes it much more modular and easier to add new variants.

avisconti commented on drivers/sensor/lps2xdf/lps2xdf.c:
```diff
@@ -0,0 +1,490 @@
+	.chip_api = &name##_chip_api,                                          \
+	IF_ENABLED(DT_NODE_HAS_COMPAT(DT_DRV_INST(inst), st_lps28dfw),         \
+		   (.fs = DT_INST_PROP(inst, fs),))                            \
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(inst, drdy_gpios),                    \
```
> Not sure if I understood this correctly. The lps22df and the lps28dfw sensor have only one interrupt pin, so one pin per device is enough here? But I could rename into int-gpios to be consistent with the other drivers.

Sure, too many devices... :D Sorry for the mistake

jremmert-phytec-iot replied:
No worries, better to double-check to not miss a bug :).

avisconti commented on drivers/sensor/lps2xdf/CMakeLists.txt:
```diff
@@ -1,12 +1,15 @@
+zephyr_library_sources(lps2xdf.c)
 zephyr_library_sources(lps22df.c)
-zephyr_library_sources_ifdef(CONFIG_LPS22DF_TRIGGER    lps22df_trigger.c)
+zephyr_library_sources(lps28dfw.c)
```
Fine, but I would use something like 
```zephyr_library_sources_ifdef(DT_HAS_LPS22DF_ENABLED, lps22df.c)```

I hope it is working

jremmert-phytec-iot replied:
It works now with `zephyr_library_sources_ifdef(CONFIG_DT_HAS_LPS22DF_ENABLED, lps22df.c)`. I like this solution, it cleaner than preprocessor macros.

I even tried this earlier without the `CONFIG_` prefix and got confused by thinking that the devicetree macros get resolved after CMake configuration.

fabiobaltieri replied:
Yeah I think the one with `CONFIG_DT_HAS_LPS22DF_ENABLED`, you wanna be careful with `DT_HAS_...` in cmake files, make sure it's behind something that can be switched off so you can still define the device in the tree and disable the driver, but I think it's fine in this case since the whole file is behind `CONFIG_LPS2XDF`. Good stuff.

avisconti reviewed (changes requested):
I love the way this code is written. Just a small nit below.
I also need to test it out on sensortile_box_pro, so I will review/approve tomorrow.
Thanks for your effort!

avisconti reviewed (approved):
It is an outstanding piece of code. Very clean! And I'm gonna re-use it for other ST sensor drivers families. Good job!!

I tested on sensortile_box_pro with lps22df.

fabiobaltieri reviewed (approved):
Good job.
